File: backend/app/routers/blocks.py
from fastapi import APIRouter, HTTPException, status, Depends, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from app.database import get_db
from app.models import User
from app.schemas import BlockCreate, BlockResponse, BlockUpdate
from app.security import get_current_user
from app.crud import blocks as crud_blocks

router = APIRouter(prefix="/blocks", tags=["blocks"])


@router.get("/", response_model=List[BlockResponse])
def list_blocks(
    skip: int = Query(0, ge=0, description="Number of blocks to skip"),
    limit: int = Query(100, ge=1, le=1000, description="Maximum number of blocks to return"),
    db: Session = Depends(get_db)
):
    """List all public blocks with pagination."""
    blocks = crud_blocks.get_blocks(db, skip=skip, limit=limit)
    return blocks


@router.get("/my", response_model=List[BlockResponse])
def list_my_blocks(
    skip: int = Query(0, ge=0, description="Number of blocks to skip"),
    limit: int = Query(100, ge=1, le=1000, description="Maximum number of blocks to return"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """List all blocks owned by the current user (including private ones)."""
    blocks = crud_blocks.get_user_blocks(db, owner_id=current_user.id, skip=skip, limit=limit)
    return blocks


@router.post("/", response_model=BlockResponse, status_code=status.HTTP_201_CREATED)
def create_block(
    block: BlockCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Create a new block."""
    try:
        db_block = crud_blocks.create_block(db=db, block=block, owner_id=current_user.id)
        return db_block
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to create block: {str(e)}"
        )


@router.get("/{block_id}", response_model=BlockResponse)
def get_block(block_id: str, db: Session = Depends(get_db)):
    """Get a specific block by ID."""
    db_block = crud_blocks.get_block(db, block_id=block_id)
    if db_block is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Block not found"
        )
    return db_block


@router.put("/{block_id}", response_model=BlockResponse)
def update_block(
    block_id: str,
    block_update: BlockUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Update a block (only the owner can update)."""
    # First check if the block exists
    db_block = crud_blocks.get_block(db, block_id=block_id)
    if db_block is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Block not found"
        )
    
    # Check if the current user is the owner
    if db_block.owner_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You can only update your own blocks"
        )
    
    # Update the block
    updated_block = crud_blocks.update_block(db, block_id=block_id, block_update=block_update)
    if updated_block is None:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Failed to update block"
        )
    
    return updated_block


@router.delete("/{block_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_block(
    block_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Delete a block (only the owner can delete)."""
    # First check if the block exists
    db_block = crud_blocks.get_block(db, block_id=block_id)
    if db_block is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Block not found"
        )
    
    # Check if the current user is the owner
    if db_block.owner_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You can only delete your own blocks"
        )
    
    # Delete the block
    deleted_block = crud_blocks.delete_block(db, block_id=block_id)
    if deleted_block is None:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Failed to delete block"
        )
    
    return None  # 204 No Content response

File: frontend/src/services/blocks.ts
import { apiService } from './api';
import { Block, BlockCreate } from './types';

class BlockService {
  // Get all public blocks with pagination and filters
  async getBlocks(params?: {
    skip?: number;
    limit?: number;
    search?: string;
    owner?: string;
  }): Promise<Block[]> {
    const queryParams = new URLSearchParams();
    
    if (params?.skip) queryParams.append('skip', params.skip.toString());
    if (params?.limit) queryParams.append('limit', params.limit.toString());
    if (params?.search) queryParams.append('search', params.search);
    if (params?.owner) queryParams.append('owner', params.owner);

    const url = `/blocks${queryParams.toString() ? `?${queryParams.toString()}` : ''}`;
    return apiService.get<Block[]>(url);
  }

  // Get user's own blocks (including private ones)
  async getMyBlocks(params?: {
    skip?: number;
    limit?: number;
  }): Promise<Block[]> {
    const queryParams = new URLSearchParams();
    
    if (params?.skip) queryParams.append('skip', params.skip.toString());
    if (params?.limit) queryParams.append('limit', params.limit.toString());

    const url = `/blocks/my${queryParams.toString() ? `?${queryParams.toString()}` : ''}`;
    return apiService.get<Block[]>(url);
  }

  // Get a specific block by ID
  async getBlock(blockId: string): Promise<Block> {
    return apiService.get<Block>(`/blocks/${blockId}`);
  }

  // Create a new block
  async createBlock(blockData: BlockCreate): Promise<Block> {
    return apiService.post<Block>('/blocks/', blockData);
  }

  // Update an existing block
  async updateBlock(blockId: string, blockData: Partial<BlockCreate>): Promise<Block> {
    return apiService.put<Block>(`/blocks/${blockId}`, blockData);
  }

  // Delete a block
  async deleteBlock(blockId: string): Promise<void> {
    return apiService.delete<void>(`/blocks/${blockId}`);
  }

  // Test a block's endpoint (optional - for future use)
  async testBlock(blockId: string, testData?: any): Promise<any> {
    return apiService.post(`/invoke/${blockId}`, {
      headers: testData?.headers,
      query_params: testData?.query_params,
      body: testData?.body,
    });
  }
}

export const blockService = new BlockService();
export default blockService;
